言及機能の実装

yama998877 opened the pull request:
日報の言及機能をアプリに追加しました。
日報の内容にすでに作成してある`http://localhost:3000/reports/:id`が含まれていると
`:id`に該当する日報に作成した日報へのリンクが表示されます。

yama998877 committed:
言及機能の実装

uriライブラリを追加。
日報のcreate,update時にcontentに`http://localhost:3000/reports/:id`のURLがあったらidを保存するように処理を追加。
上記の処理をmention_url_checkメソッドにまとめた。
update時に言及機能の保存と日報の保存が同時保存されるようにtransactionメソッドを使用。
mention.rb,report.rbに言及機能用の自己参照の多対多の関連付けを追加。
言及機能のパーシャルビューファイルを作成。
show.html.erbにパーシャルビューファイルをrender。
辞書ファイルに日報が言及された時の文言を追加。
言及機能用の中間テーブルを作成。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -51,4 +59,16 @@ def set_report
     params.require(:report).permit(:title, :content)
   end
+
+  def mention_url_check(report, str)
```
この処理は `Report` に対するものなので、 `Report` クラスにメソッドとして実装するようにしましょう。そうすると第1引数の `report` を `self` にできコードもスッキリするはずです。

yama998877 replied:
ご指摘ありがとうございます。`report.rb`にコードを移して引数を削減しました。

yuuu commented on app/views/shared/_mentions.html.erb:
```diff
@@ -0,0 +1,9 @@
+<ul>
+  <p><%= t('shared.mention.message.link') unless @report.mentioned_report_ids == [] %></p>
```
`unless @report.mentioned_report_ids == []` の部分は `if @report.mentioned_reports.exists?` のように記載するとよりシンプルですね。

yama998877 replied:
ご指摘ありがとうございます。`if @report.mentioned_reports.exists?`に修正しました。

yuuu commented on app/views/shared/_mentions.html.erb:
```diff
@@ -0,0 +1,9 @@
+<ul>
+  <p><%= t('shared.mention.message.link') unless @report.mentioned_report_ids == [] %></p>
+  <% @report.mentioned_report_ids.each do |repo_id| %>
```
ここも `@report.mentioned_reports.each` のようにすると、毎回 `find` する必要がなくなるのでおすすめです。むしろ毎回 `find` すると、ページを1回表示するだけで言及の数だけSQLが実行されるので効率が良くないですね。

yuuu replied:
ここは明示的な `ORDER BY` がされていないので、ページを表示するたびに表示順が変わってしまう懸念がありますね。

yama998877 replied:
ご指摘ありがとうございます。`@report.mentioned_reports.order(:created_at).each`に修正して`find`を削除し、実行されるSQLを削減して、作成日の昇順になるようにしました🙇‍♂️

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -51,4 +59,16 @@ def set_report
+    check_url = 'http://localhost:3000/reports/'
+    report_url_check = URI.extract(str, ['http']).uniq.select { |url| url.start_with?(check_url) }
+    repo_ids = report_url_check.map { |url| url.gsub(check_url, '').to_i }
+    existing_ids = repo_ids.select { |repo_id| Report.find_by(id: repo_id) }
```
ここも `find_by` を使うと1つの言及につき1回SQLが実行されています。1回のSQLにまとめられないか検討をお願いします 🙏

yama998877 replied:
ご指摘ありがとうございます。`Report.where(id: repo_ids).pluck(:id)`を使用するように修正して、一回のSQLにまとめました🙇‍♂️

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -51,4 +59,16 @@ def set_report
+    repo_ids = report_url_check.map { |url| url.gsub(check_url, '').to_i }
+    existing_ids = repo_ids.select { |repo_id| Report.find_by(id: repo_id) }
+
+    return existing_ids if Report.exists?(report.id)
```
更新時と新規登録時でロジックを切り替えているようですが、元々の言及を全て削除してから登録するようにすると、同じロジックで対応できそうです。

yama998877 replied:
ご指摘ありがとうございます。`mentioning_reports.destroy_all if mentioning_reports.exists?`をクラスメソッド`mention_url_check(str)`に追加して、`return`を削除して同じロジックで対応できるようにしました🙇‍♂️

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -51,4 +59,16 @@ def set_report
+
+    return existing_ids if Report.exists?(report.id)
+
+    existing_ids = existing_ids.map { |id| { mentioned_id: id } }
```
今のロジックだと、この日報自身への言及ができてしまうので、対策をお願い致します。

yama998877 replied:
ご指摘ありがとうございます。`repo_ids.delete(id)`を追加して日報自身の`id`が配列に含まれていたら配列から取り除くようにしました🙇‍♂️

yama998877 committed:
言及機能のコード修正

mention_url_checkメソッドをreport.rbに移動し、引数を削減。
update時のmention_url_checkメソッドをtransactionメソッドのブロック内に移動。
更新時の言及機能のロジックを新規作成時の言及機能のロジックを使うように修正。
日報自身に言及できないように更新時の日報のidが言及機能の処理に含まれていたら削除するように修正。
パーシャルビューファイルでmentioned_reports.exists?を使用するように修正。
無駄なSQLが実行されないようにmetioned_report_ids、Report.findを修正、削除。
言及機能の日報のリンクが作成日の昇順になるように修正。

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -21,6 +23,8 @@ def edit; end
     @report = current_user.reports.new(report_params)
 
+    str = @report.content
+    @report.mention_url_check(str)
```
このメソッド名から「メンションを作る」ということが想像しづらいので、名前の見直しをお願いします。

yuuu replied:
適切にロールバックする、という観点ではここにトランザクションは不要なのです。ただ、 `mention_url_check` と `@report.save` の間のタイミングで、別の日報が作成されて、そのURLに対してのメンションがされている場合に、メンションを取りこぼす可能性があります。

したがって排他制御という意味で、ここにはトランザクションが必要だと思いますので検討をお願いします 🙏

yuuu replied:
ちなみに、ActiveRecordのコールバックを使うと、コントローラで `content` を取り出す必要がなくなりますし、トランザクションも自動的にはってくれるのでコードがスッキリするはずです。
https://railsguides.jp/active_record_callbacks.html

yama998877 replied:
ご指摘ありがとうございます🙇‍♂️
> このメソッド名から「メンションを作る」ということが想像しづらいので、名前の見直しをお願いします。

メソッド名はメソッド内では最終的に`build`しているので`build_mention`に変更しました。

> 適切にロールバックする、という観点ではここにトランザクションは不要なのです。ただ、 mention_url_check と @report.save の間のタイミングで、別の日報が作成されて、そのURLに対してのメンションがされている場合に、メンションを取りこぼす可能性があります。
したがって排他制御という意味で、ここにはトランザクションが必要だと思いますので検討をお願いします 🙏

> ちなみに、ActiveRecordのコールバックを使うと、コントローラで content を取り出す必要がなくなりますし、トランザクションも自動的にはってくれるのでコードがスッキリするはずです。

ActiveRecordのコールバックを使って、トランザクション内で言及機能のメソッドが実行されるようにしました。
また、更新時にもコールバックを使って言及機能の更新ができそうだったので更新時のコードも修正しました🙇‍♂️

yama998877 committed:
コントローラー、モデル修正

言及機能の処理メソッド名を変更。
言及機能の処理メソッドをコールバック用のプラベートメソッドに組み込み。
保存前にコールバックメソッドが実行されるようにセット
コールバックメソッドを使うことで不要になったコントローラーのコード削除

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -1,5 +1,7 @@
 # frozen_string_literal: true
 
+require 'uri'
```
この `require` は不要になった認識なので削除しておきましょう 🙏

yama998877 replied:
ご指摘ありがとうございます。削除しました🙇‍♂️

yuuu commented on app/controllers/reports_controller.rb:
```diff
@@ -20,7 +22,6 @@ def edit; end
 
   def create
     @report = current_user.reports.new(report_params)
-
```
空行が意図せず消えているようなので、ここは戻しておきましょう。

yama998877 replied:
ご指摘ありがとうございます。戻しました🙇‍♂️

yama998877 committed:
不要なコードの削除、空行の修正

コントローラー内のrequireの削除
空行を誤って削除してしまったので修正
